src/circuit/v2/hop.ts

```diff
@@ -0,0 +1,219 @@
+}
+
+export async function handleHopProtocol (options: HopProtocolOptions): Promise<void> {
+  const { stream, request } = options
+  log('received hop message')
+  switch (request.type) {
+    case HopMessage.Type.RESERVE: await handleReserve(options); break
+    case HopMessage.Type.CONNECT: await handleConnect(options); break
+    default: {
+      log.error('invalid hop request type %s via peer %s', options.request.type, options.connection.remotePeer)
+      stream.pb(HopMessage).write({ type: HopMessage.Type.STATUS, status: Status.UNEXPECTED_MESSAGE })
+    }
+  }
+}
+
+export async function reserve (connection: Connection): Promise<Reservation> {
+  log('requesting reservation from %s', connection.remotePeer)
+  const stream = await connection.newStream([RELAY_V2_HOP_CODEC])
+  const pbstr = pbStream(stream)
+  const hopstr = pbstr.pb(HopMessage)
+  hopstr.write({ type: HopMessage.Type.RESERVE })
+
+  let response: HopMessage
+  try {
+    response = await hopstr.read()
+  } catch (e: any) {
+    log.error('error passing reserve message response from %s because', connection.remotePeer, e.message)
+    stream.close()
+    throw e
+  }
+
+  if (response.status === Status.OK && (response.reservation != null)) {
+    return response.reservation
+  }
+  const errMsg = `reservation failed with status ${response.status ?? 'undefined'}`
+  log.error(errMsg)
+  throw new Error(errMsg)
+}
+
+const isRelayAddr = (ma: Multiaddr): boolean => ma.protoCodes().includes(CIRCUIT_PROTO_CODE)
+
+async function handleReserve ({ connection, stream: pbstr, relayPeer, relayAddrs, limit, acl, reservationStore, peerStore }: HopProtocolOptions): Promise<void> {
+  const hopstr = pbstr.pb(HopMessage)
+  log('hop reserve request from %s', connection.remotePeer)
+
+  if (isRelayAddr(connection.remoteAddr)) {
+    log.error('relay reservation over circuit connection denied for peer: %p', connection.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+    return
+  }
+
+  if ((await acl?.allowReserve?.(connection.remotePeer, connection.remoteAddr)) === false) {
+    log.error('acl denied reservation to %s', connection.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+    return
+  }
+
+  const result = await reservationStore.reserve(connection.remotePeer, connection.remoteAddr)
+
+  if (result.status !== Status.OK) {
+    hopstr.write({ type: HopMessage.Type.STATUS, status: result.status })
+    return
+  }
+
+  try {
+    // tag relayed peer
+    // result.expire is non-null if `ReservationStore.reserve` returns with status == OK
+    if (result.expire != null) {
+      const ttl = new Date().getTime() - result.expire
+      await peerStore.tagPeer(relayPeer, RELAYED, { value: 1, ttl })
+        .catch((err: CodeError) => {
+          // ignore if peer is already tagged
+          if (err.code !== 'ERR_DUPLICATE_TAG') {
+            throw err
+          }
+        })
```

Comment:
If the peer is already tagged, this should update the TTL of the tag if it's going to expire before the reservation rather than swallowing the error.

Also, I can't see any tests for this?